server/resp/src/main/java/org/infinispan/server/resp/commands/list/blocking/AbstractBlockingPop.java

```diff
@@ -141,7 +141,19 @@ static CompletionStage<Collection<byte[]>> pollKeyValue(EmbeddedMultimapListCach
             });
    }
 
-   @Listener(clustered = true)
+   /**
+    * Asynchronous listener to capture events.
+    * <p>
+    * The listener captures events for entries creation and updates. The polling mechanism is triggered in case the entry
+    * is a {@link ListBucket}, that is, an object representing the RESP list to poll from.
+    * </p>
+    *
+    * <p>
+    * <b>Warning:</b> Utilize an asynchronous listener to avoid deadlock in transactional caches. Otherwise, the listener
```

Comment:
I think the issue is because the listener is called before the locks are released. So the poll operation tries to acquire the same lock the listener is holding. I didn't expect this as we use async operations. Also, this only happens in a TX cache.

Reply:
Yeah, it should be called before the locks are released. If the poll operation is acquiring a lock, we may want to offload that to the blocking thread pool then. Unfortunately, async listeners are invoked only on a single thread, to guarantee ordering between the same key, so it can be a bit of a scalability bottleneck.

Reply:
Updated. I've removed the `sync=false` and instead offload the polling method to the non-blocking executor when receiving a notification. This is only for transactional caches.

Reply:
I am not sure what is going on here then. If the issue was it was deadlocking the invoking thread when it was sync and calling the poll directly. Then we cannot use the non blocking thread pool as that means it will block the non blocking thread until the lock is released as well, which is why I recommended the blocking pool.